File: packages/typescript/src/namespaces/timer.ts
import type { OdooClientApi } from "../client-api.js";
import { formatOdooDate, parseOdooDate, parseOdooDateTime } from "../dates.js";
import { VodooError } from "../errors.js";
import type { Domain, OdooRecord } from "../types.js";

export const TIMESHEET_MODEL = "account.analytic.line";
export const TIMER_TIMER_DOMAIN: Domain = [
  ["timer_start", "!=", false],
  ["timer_pause", "=", false],
];
export const TIMER_TIMER_FIELDS = [
  "timer_start",
  "res_model",
  "res_id",
] as const;
const BASE_FIELDS = [
  "name",
  "project_id",
  "task_id",
  "unit_amount",
  "timer_start",
  "date",
] as const;

export enum TimerState {
  Running = "running",
  Stopped = "stopped",
}

export type TimerSourceKind = "task" | "ticket" | "standalone";

export class TimerSource {
  constructor(
    readonly kind: TimerSourceKind,
    readonly id: number,
    readonly name: string,
  ) {}

  get icon(): string {
    return { task: "🔧", ticket: "🎫", standalone: "⏱" }[this.kind];
  }

  get model(): string {
    return this.kind === "task"
      ? "project.task"
      : this.kind === "ticket"
        ? "helpdesk.ticket"
        : TIMESHEET_MODEL;
  }
}

export interface TimesheetObject {
  readonly id: number;
  readonly name: string;
  readonly projectName: string | null;
  readonly source: { kind: TimerSourceKind; id: number; name: string };
  readonly unitAmount: number;
  readonly state: TimerState;
  readonly elapsed: string;
  readonly date: Date;
}

export class Timesheet {
  constructor(
    readonly id: number,
    readonly name: string,
    readonly projectName: string | null,
    readonly source: TimerSource,
    readonly unitAmount: number,
    readonly timerStart: Date | null,
    readonly date: Date,
  ) {}

  get state(): TimerState {
    return this.timerStart === null ? TimerState.Stopped : TimerState.Running;
  }

  elapsedMilliseconds(now = new Date()): number {
    const base = this.unitAmount * 3_600_000;
    return this.timerStart === null
      ? base
      : base + now.getTime() - this.timerStart.getTime();
  }

  elapsedFormatted(now = new Date()): string {
    const totalSeconds = Math.trunc(this.elapsedMilliseconds(now) / 1000);
    const hours = Math.trunc(totalSeconds / 3600);
    const minutes = Math.trunc((totalSeconds % 3600) / 60);
    return `${hours}:${String(minutes).padStart(2, "0")}`;
  }

  get displayLabel(): string {
    const label =
      this.source.kind === "standalone"
        ? this.name || "Timesheet"
        : this.source.name;
    return `${this.source.icon} ${label}`;
  }

  toObject(now = new Date()): TimesheetObject {
    return {
      id: this.id,
      name: this.name,
      projectName: this.projectName,
      source: {
        kind: this.source.kind,
        id: this.source.id,
        name: this.source.name,
      },
      unitAmount: this.unitAmount,
      state: this.state,
      elapsed: this.elapsedFormatted(now),
      date: this.date,
    };
  }
}

export class TimerHandle {
  constructor(
    private readonly namespace: TimerNamespace,
    private readonly sourceKind: TimerSourceKind,
    private readonly sourceId: number,
  ) {}

  async stop(): Promise<void> {
    const active = await this.namespace.active();
    const timesheet = active.find(
      (candidate) =>
        candidate.source.kind === this.sourceKind &&
        candidate.source.id === this.sourceId,
    );
    if (timesheet !== undefined) {
      await this.namespace.stopOne(timesheet);
      return;
    }
    if (this.sourceKind === "standalone") {
      await this.namespace.stopTimesheet(this.sourceId);
      return;
    }
    throw new VodooError(
      `No running timer found for ${this.sourceKind} ${this.sourceId}`,
    );
  }
}

interface TimerBackend {
  enrichWithRunningState(
    timesheets: Timesheet[],
    uid: number,
  ): Promise<Timesheet[]>;
  startTimer(timesheet: Timesheet): Promise<void>;
  stopTimer(timesheet: Timesheet): Promise<unknown>;
}

function many2One(value: unknown): readonly [number, string] | null {
  return Array.isArray(value) &&
    typeof value[0] === "number" &&
    typeof value[1] === "string"
    ? [value[0], value[1]]
    : null;
}

function parseTimerDateTime(value: unknown): Date | null {
  if (typeof value !== "string") return null;
  try {
    return parseOdooDateTime(value);
  } catch {
    return null;
  }
}

function parseTimerDate(value: unknown): Date | null {
  if (typeof value !== "string") return null;
  try {
    return parseOdooDate(value);
  } catch {
    return null;
  }
}

function parseSource(record: OdooRecord): TimerSource {
  const task = many2One(record.task_id);
  if (task !== null) return new TimerSource("task", task[0], task[1]);
  const ticket = many2One(record.helpdesk_ticket_id);
  if (ticket !== null) return new TimerSource("ticket", ticket[0], ticket[1]);
  return new TimerSource("standalone", 0, "");
}

export function parseTimesheet(record: OdooRecord): Timesheet | null {
  if (typeof record.id !== "number") return null;
  const date = parseTimerDate(record.date);
  if (date === null) return null;
  const project = many2One(record.project_id);
  return new Timesheet(
    record.id,
    typeof record.name === "string" ? record.name : "",
    project?.[1] ?? null,
    parseSource(record),
    typeof record.unit_amount === "number" ? record.unit_amount : 0,
    parseTimerDateTime(record.timer_start),
    date,
  );
}

export function buildRunningTimer(
  record: OdooRecord,
  source: TimerSource,
  projectName: string | null,
  timerStart: Date,
  today = new Date(),
): Timesheet {
  const sourceRecordId = typeof record.id === "number" ? record.id : source.id;
  return new Timesheet(
    -sourceRecordId,
    "",
    projectName,
    source,
    0,
    timerStart,
    parseOdooDate(formatOdooDate(today)),
  );
}

export function mergeRunningTimers(
  timesheets: readonly Timesheet[],
  runningTimers: readonly Timesheet[],
): Timesheet[] {
  const result = [...timesheets];
  for (const timer of runningTimers) {
    const index = result.findIndex(
      (candidate) =>
        candidate.source.kind === timer.source.kind &&
        candidate.source.id === timer.source.id,
    );
    if (index === -1) {
      result.push(timer);
      continue;
    }
    const existing = result[index];
    if (existing === undefined) continue;
    result[index] = new Timesheet(
      existing.id,
      existing.name,
      existing.projectName,
      existing.source,
      existing.unitAmount,
      timer.timerStart,
      existing.date,
    );
  }
  return result;
}

function timerTarget(timesheet: Timesheet): readonly [string, number] {
  return timesheet.source.kind === "standalone"
    ? [TIMESHEET_MODEL, timesheet.id]
    : [timesheet.source.model, timesheet.source.id];
}

class Odoo19TimerBackend implements TimerBackend {
  constructor(private readonly client: OdooClientApi) {}

  async enrichWithRunningState(timesheets: Timesheet[]): Promise<Timesheet[]> {
    return timesheets;
  }

  async startTimer(timesheet: Timesheet): Promise<void> {
    await this.client.execute(TIMESHEET_MODEL, "action_timer_start", [
      [timesheet.id],
    ]);
  }

  stopTimer(timesheet: Timesheet): Promise<unknown> {
    return this.client.execute(TIMESHEET_MODEL, "action_timer_stop", [
      [timesheet.id],
    ]);
  }
}

class LegacyTimerBackend implements TimerBackend {
  constructor(private readonly client: OdooClientApi) {}

  async enrichWithRunningState(
    timesheets: Timesheet[],
    uid: number,
  ): Promise<Timesheet[]> {
    return mergeRunningTimers(timesheets, await this.fetchRunningTimers(uid));
  }

  async startTimer(timesheet: Timesheet): Promise<void> {
    const [model, recordId] = timerTarget(timesheet);
    await this.client.execute(model, "action_timer_start", [[recordId]]);
  }

  async stopTimer(timesheet: Timesheet): Promise<unknown> {
    const [model, recordId] = timerTarget(timesheet);
    return this.client.execute(model, "action_timer_stop", [[recordId]]);
  }

  private async fetchRunningTimers(uid: number): Promise<Timesheet[]> {
    let records: OdooRecord[];
    try {
      records = await this.client.searchRead("timer.timer", {
        domain: [["user_id", "=", uid], ...TIMER_TIMER_DOMAIN],
        fields: TIMER_TIMER_FIELDS,
      });
    } catch {
      return [];
    }
    const result: Timesheet[] = [];
    for (const record of records) {
      if (
        typeof record.res_model !== "string" ||
        typeof record.res_id !== "number"
      ) {
        continue;
      }
      const timerStart = parseTimerDateTime(record.timer_start);
      if (timerStart === null) continue;
      if (record.res_model === "project.task") {
        let name = `Task #${record.res_id}`;
        let projectName: string | null = null;
        try {
          const tasks = await this.client.searchRead("project.task", {
            domain: [["id", "=", record.res_id]],
            fields: ["display_name", "project_id"],
            limit: 1,
          });
          if (typeof tasks[0]?.display_name === "string")
            name = tasks[0].display_name;
          projectName = many2One(tasks[0]?.project_id)?.[1] ?? null;
        } catch {
          // Keep stable fallback labels when the source cannot be read.
        }
        result.push(
          buildRunningTimer(
            record,
            new TimerSource("task", record.res_id, name),
            projectName,
            timerStart,
          ),
        );
      } else if (record.res_model === "helpdesk.ticket") {
        let name = `Ticket #${record.res_id}`;
        try {
          const tickets = await this.client.searchRead("helpdesk.ticket", {
            domain: [["id", "=", record.res_id]],
            fields: ["display_name"],
            limit: 1,
          });
          if (typeof tickets[0]?.display_name === "string")
            name = tickets[0].display_name;
        } catch {
          // Keep stable fallback labels when the source cannot be read.
        }
        result.push(
          buildRunningTimer(
            record,
            new TimerSource("ticket", record.res_id, name),
            null,
            timerStart,
          ),
        );
      }
    }
    return result;
  }
}

interface StopWizard {
  readonly model: string;
  readonly values: Readonly<Record<string, unknown>>;
  readonly method: string;
  readonly context: Readonly<Record<string, unknown>>;
}

function stopWizard(result: unknown): StopWizard | null {
  if (typeof result !== "object" || result === null) return null;
  const action = result as Record<string, unknown>;
  if (
    action.type !== "ir.actions.act_window" ||
    typeof action.res_model !== "string"
  ) {
    return null;
  }
  const context =
    typeof action.context === "object" && action.context !== null
      ? (action.context as Readonly<Record<string, unknown>>)
      : {};
  if (action.res_model === "project.task.create.timesheet") {
    return {
      model: action.res_model,
      values: {
        task_id: context.active_id ?? 0,
        description: "/",
        time_spent: context.default_time_spent ?? 0,
      },
      method: "save_timesheet",
      context,
    };
  }
  if (action.res_model === "helpdesk.ticket.create.timesheet") {
    return {
      model: action.res_model,
      values: {
        ticket_id: context.active_id ?? 0,
        description: "/",
        time_spent: context.default_time_spent ?? 0,
      },
      method: "action_generate_timesheet",
      context,
    };
  }
  if (action.res_model === "hr.timesheet.stop.timer.confirmation.wizard") {
    return {
      model: action.res_model,
      values: { timesheet_id: context.default_timesheet_id ?? 0 },
      method: "action_stop_timer",
      context,
    };
  }
  return null;
}

export interface TimerListOptions {
  days?: number;
  limit?: number | null;
}

export class TimerNamespace {
  private helpdeskField: boolean | undefined;

  constructor(private readonly client: OdooClientApi) {}

  async list(options: TimerListOptions = {}): Promise<Timesheet[]> {
    const uid = await this.client.getUid();
    const fields = await this.fields();
    const days = options.days ?? 0;
    const domain: Array<Domain[number]> = [["user_id", "=", uid]];
    if (days >= 0) {
      domain.push([
        "date",
        ">=",
        formatOdooDate(new Date(Date.now() - days * 86_400_000)),
      ]);
    }
    const records = await this.client.searchRead(TIMESHEET_MODEL, {
      domain,
      fields,
      order: "date desc",
      limit: options.limit ?? null,
    });
    const timesheets = records
      .map(parseTimesheet)
      .filter((timesheet): timesheet is Timesheet => timesheet !== null);
    return this.backend().enrichWithRunningState(timesheets, uid);
  }

  async active(): Promise<Timesheet[]> {
    const uid = await this.client.getUid();
    if (!this.client.isJson2) {
      return (await this.list()).filter(
        (timesheet) => timesheet.timerStart !== null,
      );
    }
    const records = await this.client.searchRead(TIMESHEET_MODEL, {
      domain: [
        ["user_id", "=", uid],
        ["timer_start", "!=", false],
      ],
      fields: await this.fields(),
      order: "date desc",
    });
    return records
      .map(parseTimesheet)
      .filter(
        (timesheet): timesheet is Timesheet =>
          timesheet !== null && timesheet.timerStart !== null,
      );
  }

  async startTask(taskId: number): Promise<TimerHandle> {
    await this.client.execute("project.task", "action_timer_start", [[taskId]]);
    return new TimerHandle(this, "task", taskId);
  }

  async startTicket(ticketId: number): Promise<TimerHandle> {
    await this.client.execute("helpdesk.ticket", "action_timer_start", [
      [ticketId],
    ]);
    return new TimerHandle(this, "ticket", ticketId);
  }

  async startTimesheet(timesheetId: number): Promise<TimerHandle> {
    const timesheet = await this.loadTimesheet(timesheetId);
    await this.backend().startTimer(timesheet);
    const sourceId =
      timesheet.source.kind === "standalone"
        ? timesheetId
        : timesheet.source.id;
    return new TimerHandle(this, timesheet.source.kind, sourceId);
  }

  async stopTimesheet(timesheetId: number): Promise<void> {
    const timesheet = await this.loadTimesheet(timesheetId);
    await this.handleStopWizard(await this.backend().stopTimer(timesheet));
  }

  async stopOne(timesheet: Timesheet): Promise<void> {
    await this.client.getUid();
    await this.handleStopWizard(await this.backend().stopTimer(timesheet));
  }

  async stop(): Promise<Timesheet[]> {
    const active = await this.active();
    const backend = this.backend();
    for (const timesheet of active) {
      await this.handleStopWizard(await backend.stopTimer(timesheet));
    }
    return active;
  }

  private backend(): TimerBackend {
    return this.client.isJson2
      ? new Odoo19TimerBackend(this.client)
      : new LegacyTimerBackend(this.client);
  }

  private async hasHelpdeskField(): Promise<boolean> {
    if (this.helpdeskField !== undefined) return this.helpdeskField;
    try {
      await this.client.searchRead(TIMESHEET_MODEL, {
        domain: [],
        fields: ["id", "helpdesk_ticket_id"],
        limit: 1,
      });
      this.helpdeskField = true;
    } catch {
      this.helpdeskField = false;
    }
    return this.helpdeskField;
  }

  private async fields(): Promise<string[]> {
    const fields: string[] = [...BASE_FIELDS];
    if (await this.hasHelpdeskField()) fields.push("helpdesk_ticket_id");
    return fields;
  }

  private async loadTimesheet(timesheetId: number): Promise<Timesheet> {
    const records = await this.client.searchRead(TIMESHEET_MODEL, {
      domain: [["id", "=", timesheetId]],
      fields: await this.fields(),
      limit: 1,
    });
    if (records[0] === undefined)
      throw new VodooError(`Timesheet ${timesheetId} not found`);
    const timesheet = parseTimesheet(records[0]);
    if (timesheet === null)
      throw new VodooError(`Failed to parse timesheet ${timesheetId}`);
    return timesheet;
  }

  private async handleStopWizard(result: unknown): Promise<void> {
    const wizard = stopWizard(result);
    if (wizard === null) return;
    const wizardId = await this.client.create(wizard.model, wizard.values);
    await this.client.execute(wizard.model, wizard.method, [[wizardId]], {
      context: wizard.context,
    });
  }
}

File: Sources/Vodoo/Timer.swift
import Foundation

public let timesheetModel = "account.analytic.line"
public let timerTimerDomain: Domain = [
    .array([.string("timer_start"), .string("!="), .bool(false)]),
    .array([.string("timer_pause"), .string("="), .bool(false)]),
]

public enum TimerState: String, Sendable, Equatable { case running, stopped }
public enum TimerSourceKind: String, Sendable, Equatable { case task, ticket, standalone }

public struct TimerSource: Sendable, Equatable {
    public let kind: TimerSourceKind
    public let id: Int
    public let name: String

    public init(kind: TimerSourceKind, id: Int, name: String) {
        self.kind = kind
        self.id = id
        self.name = name
    }

    public var icon: String {
        switch kind { case .task: "🔧"; case .ticket: "🎫"; case .standalone: "⏱" }
    }

    public var model: String {
        switch kind {
        case .task: "project.task"
        case .ticket: "helpdesk.ticket"
        case .standalone: timesheetModel
        }
    }
}

public struct Timesheet: Sendable, Equatable {
    public let id: Int
    public let name: String
    public let projectName: String?
    public let source: TimerSource
    public let unitAmount: Double
    public let timerStart: Date?
    public let date: Date

    public init(
        id: Int, name: String, projectName: String?, source: TimerSource,
        unitAmount: Double, timerStart: Date?, date: Date
    ) {
        self.id = id
        self.name = name
        self.projectName = projectName
        self.source = source
        self.unitAmount = unitAmount
        self.timerStart = timerStart
        self.date = date
    }

    public var state: TimerState { timerStart == nil ? .stopped : .running }

    public func elapsed(now: Date = Date()) -> TimeInterval {
        unitAmount * 3_600 + (timerStart.map { now.timeIntervalSince($0) } ?? 0)
    }

    public func elapsedFormatted(now: Date = Date()) -> String {
        let total = Int(elapsed(now: now))
        return String(format: "%d:%02d", total / 3_600, (total % 3_600) / 60)
    }

    public var displayLabel: String {
        let label = source.kind == .standalone ? (name.isEmpty ? "Timesheet" : name) : source.name
        return "\(source.icon) \(label)"
    }
}

public final class TimerHandle: @unchecked Sendable {
    private let namespace: TimerNamespace
    private let sourceKind: TimerSourceKind
    private let sourceID: Int

    init(namespace: TimerNamespace, sourceKind: TimerSourceKind, sourceID: Int) {
        self.namespace = namespace
        self.sourceKind = sourceKind
        self.sourceID = sourceID
    }

    public func stop() async throws {
        if let timesheet = try await namespace.active().first(where: {
            $0.source.kind == sourceKind && $0.source.id == sourceID
        }) {
            try await namespace.stop(timesheet)
            return
        }
        if sourceKind == .standalone {
            try await namespace.stopTimesheet(sourceID)
            return
        }
        throw VodooError.operation("No running timer found for \(sourceKind.rawValue) \(sourceID)")
    }
}

public final class TimerNamespace: @unchecked Sendable {
    private let client: any OdooClientAPI
    private var helpdeskField: Bool?

    public init(client: any OdooClientAPI) { self.client = client }

    public func list(days: Int = 0, limit: Int? = nil) async throws -> [Timesheet] {
        let uid = try await client.getUID()
        var domain: Domain = [
            .array([.string("user_id"), .string("="), .integer(uid)]),
        ]
        if days >= 0 {
            domain.append(.array([
                .string("date"), .string(">="),
                .string(try OdooDateCodec.formatDate(Date().addingTimeInterval(Double(-days) * 86_400))),
            ]))
        }
        let records = try await client.searchRead(
            model: timesheetModel, domain: domain, fields: await fields(),
            limit: limit, offset: 0, order: "date desc"
        )
        let timesheets = records.compactMap(parseTimesheet)
        if try await client.transportDialect() == .json2 { return timesheets }
        return mergeRunningTimers(timesheets, try await fetchRunningTimers(uid: uid))
    }

    public func active() async throws -> [Timesheet] {
        let days = try await client.transportDialect() == .json2 ? -1 : 0
        return try await list(days: days).filter { $0.timerStart != nil }
    }

    public func startTask(_ taskID: Int) async throws -> TimerHandle {
        _ = try await client.execute(
            model: "project.task", method: "action_timer_start",
            args: [.array([.integer(taskID)])], kwargs: nil
        )
        return TimerHandle(namespace: self, sourceKind: .task, sourceID: taskID)
    }

    public func startTicket(_ ticketID: Int) async throws -> TimerHandle {
        _ = try await client.execute(
            model: "helpdesk.ticket", method: "action_timer_start",
            args: [.array([.integer(ticketID)])], kwargs: nil
        )
        return TimerHandle(namespace: self, sourceKind: .ticket, sourceID: ticketID)
    }

    public func startTimesheet(_ timesheetID: Int) async throws -> TimerHandle {
        let timesheet = try await loadTimesheet(timesheetID)
        if try await client.transportDialect() == .json2 {
            _ = try await client.execute(
                model: timesheetModel, method: "action_timer_start",
                args: [.array([.integer(timesheet.id)])], kwargs: nil
            )
        } else {
            let target = timerTarget(timesheet)
            _ = try await client.execute(
                model: target.0, method: "action_timer_start",
                args: [.array([.integer(target.1)])], kwargs: nil
            )
        }
        let sourceID = timesheet.source.kind == .standalone ? timesheetID : timesheet.source.id
        return TimerHandle(namespace: self, sourceKind: timesheet.source.kind, sourceID: sourceID)
    }

    public func stopTimesheet(_ timesheetID: Int) async throws {
        try await stop(loadTimesheet(timesheetID))
    }

    public func stop(_ timesheet: Timesheet) async throws {
        _ = try await client.getUID()
        let result: JSONValue
        if try await client.transportDialect() == .json2 {
            result = try await client.execute(
                model: timesheetModel, method: "action_timer_stop",
                args: [.array([.integer(timesheet.id)])], kwargs: nil
            )
        } else {
            let target = timerTarget(timesheet)
            result = try await client.execute(
                model: target.0, method: "action_timer_stop",
                args: [.array([.integer(target.1)])], kwargs: nil
            )
        }
        try await handleStopWizard(result)
    }

    @discardableResult
    public func stopAll() async throws -> [Timesheet] {
        let values = try await active()
        for timesheet in values { try await stop(timesheet) }
        return values
    }

    private func hasHelpdeskField() async -> Bool {
        if let helpdeskField { return helpdeskField }
        do {
            _ = try await client.searchRead(
                model: timesheetModel, domain: [], fields: ["id", "helpdesk_ticket_id"],
                limit: 1, offset: 0, order: nil
            )
            helpdeskField = true
        } catch {
            helpdeskField = false
        }
        return helpdeskField ?? false
    }

    private func fields() async -> [String] {
        var values = ["name", "project_id", "task_id", "unit_amount", "timer_start", "date"]
        if await hasHelpdeskField() { values.append("helpdesk_ticket_id") }
        return values
    }

    private func loadTimesheet(_ id: Int) async throws -> Timesheet {
        let records = try await client.searchRead(
            model: timesheetModel,
            domain: [.array([.string("id"), .string("="), .integer(id)])],
            fields: await fields(), limit: 1, offset: 0, order: nil
        )
        guard let record = records.first else { throw VodooError.operation("Timesheet \(id) not found") }
        guard let result = parseTimesheet(record) else {
            throw VodooError.operation("Failed to parse timesheet \(id)")
        }
        return result
    }

    private func fetchRunningTimers(uid: Int) async throws -> [Timesheet] {
        let records: [OdooRecord]
        do {
            records = try await client.searchRead(
                model: "timer.timer",
                domain: [.array([.string("user_id"), .string("="), .integer(uid)])] + timerTimerDomain,
                fields: ["timer_start", "res_model", "res_id"],
                limit: nil, offset: 0, order: nil
            )
        } catch { return [] }
        var result: [Timesheet] = []
        for record in records {
            guard let model = record["res_model"]?.stringValue,
                  let id = record["res_id"]?.intValue,
                  let startText = record["timer_start"]?.stringValue,
                  let timerStart = try? OdooDateCodec.parseDateTime(startText) else { continue }
            if model == "project.task" {
                var name = "Task #\(id)"
                var project: String?
                if let task = try? await client.searchRead(
                    model: model,
                    domain: [.array([.string("id"), .string("="), .integer(id)])],
                    fields: ["display_name", "project_id"], limit: 1, offset: 0, order: nil
                ).first {
                    name = task["display_name"]?.stringValue ?? name
                    project = relationName(task["project_id"])
                }
                result.append(buildRunningTimer(
                    record: record, source: TimerSource(kind: .task, id: id, name: name),
                    projectName: project, timerStart: timerStart
                ))
            } else if model == "helpdesk.ticket" {
                var name = "Ticket #\(id)"
                if let ticket = try? await client.searchRead(
                    model: model,
                    domain: [.array([.string("id"), .string("="), .integer(id)])],
                    fields: ["display_name"], limit: 1, offset: 0, order: nil
                ).first {
                    name = ticket["display_name"]?.stringValue ?? name
                }
                result.append(buildRunningTimer(
                    record: record, source: TimerSource(kind: .ticket, id: id, name: name),
                    projectName: nil, timerStart: timerStart
                ))
            }
        }
        return result
    }

    private func handleStopWizard(_ result: JSONValue) async throws {
        guard let action = result.objectValue,
              action["type"]?.stringValue == "ir.actions.act_window",
              let model = action["res_model"]?.stringValue else { return }
        let context = action["context"]?.objectValue ?? [:]
        let values: OdooRecord
        let method: String
        switch model {
        case "project.task.create.timesheet":
            values = [
                "task_id": context["active_id"] ?? .integer(0),
                "description": .string("/"),
                "time_spent": context["default_time_spent"] ?? .integer(0),
            ]
            method = "save_timesheet"
        case "helpdesk.ticket.create.timesheet":
            values = [
                "ticket_id": context["active_id"] ?? .integer(0),
                "description": .string("/"),
                "time_spent": context["default_time_spent"] ?? .integer(0),
            ]
            method = "action_generate_timesheet"
        case "hr.timesheet.stop.timer.confirmation.wizard":
            values = ["timesheet_id": context["default_timesheet_id"] ?? .integer(0)]
            method = "action_stop_timer"
        default: return
        }
        let wizardID = try await client.create(model: model, values: values, context: nil)
        _ = try await client.execute(
            model: model, method: method, args: [.array([.integer(wizardID)])],
            kwargs: ["context": .object(context)]
        )
    }
}

public func parseTimesheet(_ record: OdooRecord) -> Timesheet? {
    guard let id = record["id"]?.intValue,
          let dateText = record["date"]?.stringValue,
          let date = try? OdooDateCodec.parseDate(dateText) else { return nil }
    let source: TimerSource
    if let task = relationID(record["task_id"]) {
        source = TimerSource(kind: .task, id: task, name: relationName(record["task_id"]) ?? "")
    } else if let ticket = relationID(record["helpdesk_ticket_id"]) {
        source = TimerSource(
            kind: .ticket, id: ticket, name: relationName(record["helpdesk_ticket_id"]) ?? ""
        )
    } else {
        source = TimerSource(kind: .standalone, id: 0, name: "")
    }
    let timerStart = record["timer_start"]?.stringValue.flatMap { try? OdooDateCodec.parseDateTime($0) }
    return Timesheet(
        id: id, name: record["name"]?.stringValue ?? "",
        projectName: relationName(record["project_id"]), source: source,
        unitAmount: record["unit_amount"]?.doubleValue ?? 0,
        timerStart: timerStart, date: date
    )
}

public func buildRunningTimer(
    record: OdooRecord,
    source: TimerSource,
    projectName: String?,
    timerStart: Date,
    today: Date = Date()
) -> Timesheet {
    var calendar = Calendar(identifier: .gregorian)
    calendar.timeZone = TimeZone(secondsFromGMT: 0)!
    return Timesheet(
        id: -(record["id"]?.intValue ?? source.id), name: "", projectName: projectName,
        source: source, unitAmount: 0, timerStart: timerStart, date: calendar.startOfDay(for: today)
    )
}

public func mergeRunningTimers(_ timesheets: [Timesheet], _ running: [Timesheet]) -> [Timesheet] {
    var result = timesheets
    for timer in running {
        if let index = result.firstIndex(where: {
            $0.source.kind == timer.source.kind && $0.source.id == timer.source.id
        }) {
            let current = result[index]
            result[index] = Timesheet(
                id: current.id, name: current.name, projectName: current.projectName,
                source: current.source, unitAmount: current.unitAmount,
                timerStart: timer.timerStart, date: current.date
            )
        } else {
            result.append(timer)
        }
    }
    return result
}

private func timerTarget(_ timesheet: Timesheet) -> (String, Int) {
    timesheet.source.kind == .standalone
        ? (timesheetModel, timesheet.id) : (timesheet.source.model, timesheet.source.id)
}
